### PR TITLE
faster maximum, minimum (#28849)

### PR DESCRIPTION
**Pull request by jw3126**
Borrowing an idea from @antoine-levitt https://github.com/JuliaLang/julia/issues/26003#issuecomment-364907362.
~There is one slight difference to the current `maximum/minimum` behavior. Namely `-0.0` and `0.0` are not always ordered correctly (the same thing happens with numpy.):~

## Old
```julia
julia> using BenchmarkTools; arr = randn(10^6); @benchmark maximum($arr)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     5.818 ms (0.00% GC)
  median time:      6.048 ms (0.00% GC)
  mean time:        6.054 ms (0.00% GC)
  maximum time:     6.973 ms (0.00% GC)
  --------------
  samples:          825
  evals/sample:     1
```
## Numpy
```python
In [1]: import numpy as np

In [2]: arr = np.random.randn(10**6);

In [3]: %timeit np.max(arr)
429 µs ± 2.25 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

## ~New~
```julia
julia> using BenchmarkTools; arr = randn(10^6); @benchmark maximum($arr)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     1.215 ms (0.00% GC)
  median time:      1.264 ms (0.00% GC)
  mean time:        1.265 ms (0.00% GC)
  maximum time:     1.420 ms (0.00% GC)
  --------------
  samples:          3930
  evals/sample:     1
```

## ~New~
```julia
julia> using BenchmarkTools; arr = randn(10^6); @benchmark maximum($arr)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     909.563 μs (0.00% GC)
  median time:      937.093 μs (0.00% GC)
  mean time:        947.286 μs (0.00% GC)
  maximum time:     1.398 ms (0.00% GC)
  --------------
  samples:          5234
  evals/sample:     1
```
## New
```julia
julia> using BenchmarkTools; arr = randn(10^6); @benchmark maximum($arr)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     566.788 μs (0.00% GC)
  median time:      582.718 μs (0.00% GC)
  mean time:        590.280 μs (0.00% GC)
  maximum time:     880.178 μs (0.00% GC)
  --------------
  samples:          8369
  evals/sample:     1

```

### CODE REVIEW

File: base/reduce.jl [hunk 1]
```diff
@@ -453,16 +453,23 @@ prod(a) = mapreduce(identity, mul_prod, a)
...
                         A::AbstractArray, first::Int, last::Int)
-    # locate the first non NaN number
+    op_fast = _fast(op)
     @inbounds a1 = A[first]
     v = mapreduce_first(f, op, a1)
-    i = first + 1
-    while (v == v) && (i <= last)
+    for i in (first+1):last
+        # short circuit NaN correctly
+        (v == v) || return v
```
Comment (jw3126) [inline base/reduce.jl]: It is not necessary to check this every iteration. Eliding that check gives another 30% speedup.

File: base/reduce.jl [hunk 2]
```diff
@@ -453,16 +453,23 @@ prod(a) = mapreduce(identity, mul_prod, a)
 
 ## maximum & minimum
 
+# these propagate NaN correctly only in the second argument
+_max_fast(x, y) = ifelse(x > y, x, y)
```
Comment (KristofferC) [inline base/reduce.jl]: Could perhaps just define

```jl
_op_fast(::typeof(max), x, y) = ifelse(x > y, x, y)
_op_fast(::typeof(min), x, y) = ifelse(x < y, x, y)
```

and just use _`_op_fast(op, v, f(ai)` in `mapreduce_impl`?

Comment (jw3126) [reply base/reduce.jl]: I like this idea.

**Review by StefanKarpinski — APPROVED** [review]

**Comment by chethega** [comment]
Looking at `@code_native`, it seems that vectorization fails?
```
julia> _fmax(x,y) = ifelse(x>y, x, y);
julia> function _fmaximum(arr)
              v=arr[1]
              @simd for i=1:length(arr)
              @inbounds v = _fmax(v,arr[i])
              end
              v
              end
julia> function _fmaximum2(arr)
              v1 = arr[1]
              v2 = arr[1]
              v3 = arr[1]
              v4=arr[1]
              isnan_ = isnan(v1) | isnan(v2)
              @simd for i=1:4:length(arr)
              @inbounds v1 = _fmax(v1,arr[i])
              @inbounds v2 = _fmax(v1,arr[i+1])
              @inbounds v3 = _fmax(v1,arr[i+2])
              @inbounds v4 = _fmax(v1,arr[i+3])
              isnan_ |= isnan(v1) | isnan(v2) | isnan(v3) | isnan(v4)
              end
              !isnan_ && return convert(typeof(v1),NaN)
              max(v1,v2,v3,v4)
              end

julia> function _fmaximum3(arr)
              v1 = arr[1]
              v2 = arr[1]
              v3 = arr[1]
              v4=arr[1]
              #isnan_ = isnan(v1) | isnan(v2)
              @simd for i=1:4:length(arr)
              @inbounds v1 = _fmax(v1,arr[i])
              @inbounds v2 = _fmax(v1,arr[i+1])
              @inbounds v3 = _fmax(v1,arr[i+2])
              @inbounds v4 = _fmax(v1,arr[i+3])
              #isnan_ |= isnan(v1) | isnan(v2) | isnan(v3) | isnan(v4)
              end
              #!isnan_ && return convert(typeof(v1),NaN)
              max(v1,v2,v3,v4)
              end


julia> using BenchmarkTools
julia> arr = randn(10^6);
julia> @btime maximum($arr);
  7.640 ms (0 allocations: 0 bytes)

julia> @btime _fmaximum($arr);
  1.341 ms (0 allocations: 0 bytes)

julia> @btime _fmaximum2($arr);
  1.021 ms (0 allocations: 0 bytes)

julia> @btime _fmaximum3($arr);
  537.057 μs (0 allocations: 0 bytes)

julia> arr = randn(Float32,10^6);

julia> @btime maximum($arr);
  7.233 ms (0 allocations: 0 bytes)

julia> @btime _fmaximum($arr);
  1.323 ms (0 allocations: 0 bytes)

julia> @btime _fmaximum2($arr);
  1.003 ms (0 allocations: 0 bytes)

julia> @btime _fmaximum3($arr);
  340.281 μs (0 allocations: 0 bytes)
```
Can we get properly vectorized (SIMD) code here? Does LLVM understand that min/max are associative?

**Comment by jw3126** [comment]
Oh nice @chethega. The timing of `_fmaximum3` looks like that of numpy. So I guess it uses SIMD? Should I update the PR with such a code? Or is this a compiler glitch, thats easy to fix?

**Comment by chethega** [comment]
I don't know. Your PR gives a nice 5x boost, but there is another big factor on the table. 

I guess some llvm expert needs to chime in on associativity of min/max algebra; did we write this in a bad way, are we missing an optimization pass or is this a problem of llvm? Or is this just stuck [upstream](https://reviews.llvm.org/rL345220) and will solve itself soon? 

If the latter, then I'd say merge and keep an issue open to revisit this later, once the fundamentals in llvm are ready. Nobody got grey hair waiting for the slow `maximum` yet, so people can enjoy the 5x speedup until they get the next 4x; salami tactics yay. 

Second point: I don't understand why your implementation is correct with `NaN`. 
```
julia> fm(x,y)=ifelse(x > y, x, y)
julia> A=[1.0, NaN, 2.0];
julia> fm(fm(A[1],A[2]), A[3])
2.0
```
Does your implementation handle that correctly? Regardless, could you add that test-case, just for my piece of mind? I think this is an example that any future PR can plausibly get wrong, which makes it a good test (even better test: `[1.0, 2.0, NaN, 4.0, 5.0]`).

**Comment by jw3126** [comment]
Yes the implentation should be `NaN` correcect, there are also tests. The trick is to ensure, that all `NaN` are either the second argument or checked explicitly.
```julia
julia> fm(x,y)=ifelse(x > y, x, y)
fm (generic function with 1 method)

julia> A=[1.0, NaN, 2.0];

julia> fm(fm(A[1],A[2]), A[3])
2.0

julia> fm(A[3], fm(A[1],A[2]))
NaN
```

**Comment by chethega** [comment]
My problem is that I don't see how you ensure that. If you fold a `NaN` into a number, you get a `NaN`; good. But if you fold a number into a `NaN`, then you get the number, and since you only check every `chunk_len`, one of the `NaN`s could escape.

```
julia> @eval Base _fast(::typeof(max), x, y) = ifelse(x > y, x, y)
julia> @eval Base _fast(::typeof(min), x, y) = ifelse(x < y, x, y)

julia> @eval Base function mapreduce_impl(f::ft, op::Union{typeof(max), typeof(min)},A::AbstractArray, first::Int, last::Int) where ft
        a1 = @inbounds A[first]
        v = mapreduce_first(f, op, a1)
        chunk_len = 256
        stop = first
        while last - stop > 0
        v == v || return v
        start = stop + 1
        stop = min(stop + chunk_len, last)
        for i in start:stop
        @inbounds ai = A[i]
        v = _fast(op, v, f(ai))
        end
        end
        v
        end

julia> A=rand(5); A[3]=NaN; maximum(A)
NaN
julia> A=rand(1000); A[400]=NaN; maximum(A)
0.9991780990981842
```

**Comment by jw3126** [comment]
You are right @chethega the current state of the PR is incorrect.

**Comment by chethega** [comment]
That looks correct, but gives almost a 2x slowdown compared to the original PR :(

Somewhat faster variant for floats on my machine is
```
julia> function _maximum(a::AbstractArray{T}) where T
       v0=typemin(T)
       i1 = false
       rs = mapreduce(_lift, _red, a; init=(v0,i1))
       rs[2] && return T(NaN)
       rs[1]
       end
julia> fmax(x,y)=ifelse(x > y, x, y)
julia> _lift(x)=(x, isnan(x));
julia> _red(a,b)=( (v1,i1) = a; (v2,i2)=b; (fmax(v1,v2),i1|i2) )
julia> @btime _maximum($arr)
  1.621 ms (0 allocations: 0 bytes)
```
I think this is because (1) the above `fmax` is a hardware instruction, and (2) we are latency constrained. That is, apart from the maybe upstream issue of using SIMD / vector units, we have all these nice execution ports idling around, because the code is written such that we have one giant dependency chain. The above variant breaks at least the nan-check out of that dependency chain.

**Comment by jw3126** [comment]
That is weird, on my machine, it is the same speed as original PR.

**Comment by chethega** [comment]
Nice, no more objections from me. We might want a `@fastmath` as well, for skipping nan-handling and signed zeros. I think passing an `Ordering` is a cool idea but should be a separate feature PR.

Can we get a nanosoldier run? Just in case of surprises?

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/30320#issuecomment-446386177) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/8f8a11eac303ea012023e97921813acda0546400/39cf3c2_vs_caa0273/report.md). cc @ararslan

**Comment by jw3126** [comment]
Any ideas why CI fails?

**Comment by fredrikekre** [comment]
> Any ideas why CI fails?

 #30356

**Comment by jw3126** [comment]
Can somebody rerun appveyor?

**Comment by jw3126** [comment]
Any objections left?

File: base/reduce.jl [hunk 3]
```diff
@@ -452,19 +452,63 @@ julia> prod(1:20)
...
-    i = first + 1
-    while (v == v) && (i <= last)
+    a1 = @inbounds A[first]
+    v1 = mapreduce_first(f, op, a1)
+    v2 = v3 = v4 = v1
+    chunk_len = 256
+    start = first
+    stop  = start + chunk_len - 4
+    while stop <= last
+        isnan(v1) && return v1
```
Comment (nalimilan) [inline base/reduce.jl]: This broke `maximum` on non-numeric values. This is visible only with long arrays:
```julia
julia> maximum(rand(Char, 500))
ERROR: MethodError: no method matching isnan(::Char)
Closest candidates are:
  isnan(::BigFloat) at mpfr.jl:879
  isnan(::Missing) at missing.jl:83
  isnan(::Float16) at float.jl:530
  ...
Stacktrace:
 [1] mapreduce_impl(::typeof(identity), ::typeof(max), ::Array{Char,1}, ::Int64, ::Int64) at ./reduce.jl:482
 [2] _mapreduce at ./reduce.jl:320 [inlined]
 [3] _mapreduce_dim at ./reducedim.jl:308 [inlined]
 [4] #mapreduce#549 at ./reducedim.jl:304 [inlined]
 [5] mapreduce at ./reducedim.jl:304 [inlined]
 [6] _maximum at ./reducedim.jl:653 [inlined]
 [7] _maximum at ./reducedim.jl:652 [inlined]
 [8] #maximum#555 at ./reducedim.jl:648 [inlined]
 [9] maximum(::Array{Char,1}) at ./reducedim.jl:648
 [10] top-level scope at none:0
```

**Comment by KristofferC** [comment]
Right now we are only backporting bugfixes to 1.1.0. This can go in 1.1.1.

File: base/reduce.jl [hunk 4]
```diff
@@ -452,19 +452,63 @@ julia> prod(1:20)
...
+        isnan(v2) && return v2
+        isnan(v3) && return v3
+        isnan(v4) && return v4
+        @inbounds for i in start:4:stop
+            v1 = _fast(op, v1, f(A[i+1]))
```
Comment (ararslan) [inline base/reduce.jl]: Won't this break support for arrays with strides other than 1?

**Comment by ViralBShah** [comment]
Reverting due to the reported breakage. Probably should also be added to tests.

**Pull request merged**